Sources/SwiftDocCTestUtilities/XCTestCase+TemporaryDirectory.swift

```diff
@@ -0,0 +1,55 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2021 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See https://swift.org/LICENSE.txt for license information
+ See https://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import Foundation
+import XCTest
+
+// These helpers methods exist to put temp files for different test executions in different locations when running in Swift CI.
+
+public extension XCTestCase {
+    
+    @available(*, deprecated, message: "Use `createTemporaryDirectory` instead in unit tests to avoid referencing a shared location in Swift CI.")
+    func NSTemporaryDirectory() -> String {
+        return Foundation.NSTemporaryDirectory()
+    }
+    
+    /// Creates a new temporary directory and returns the URL of that directory.
+    ///
+    /// After the current test method has returned the temporary directory is automatically removed.
+    ///
+    /// - Parameters:
+    ///   - pathComponents: Additional path components to add to the temporary URL.
+    ///   - fileManager: The file manager that will create the directory.
+    /// - Returns: The URL of the newly created directory.
+    func createTemporaryDirectory(pathComponents: String..., fileManager: FileManager = .default) throws -> URL {
+        let bundleParentDir = Bundle(for: Self.self).bundleURL.deletingLastPathComponent()
+        let baseURL = bundleParentDir.appendingPathComponent(name.replacingWhitespaceAndPunctuation(with: "-"))
```

Comment:
Should we consider creating a parent `TemporarySwiftDocCTestDirectory-\(some-static-unique-process-identifier)` to place the individual test directories based on the test name in? Otherwise I think it's possible we'd run into collisions in the future.

I'm imagining something like:

```swift
// Define this statically so that we don't generate a new unique base directory
// every time we create a temporary directory
private static let uniqueProcessIdentifier = ProcessInfo.processInfo.globallyUniqueString
```

```suggestion
        let baseURL = bundleParentDir
            .appendingPathComponent(
                "TemporarySwiftDocCTestDirectory-\(uniqueProcessIdentifier)", 
                isDirectory: true
             )
            .appendingPathComponent(
                name.replacingWhitespaceAndPunctuation(with: "-"),
                isDirectory: true
            )
```

Reply:
Collisions shouldn't be an issue here. 

The `bundleParentDir` is a path inside the build directory for the specific checkout, so that should be unique if Swift CI is testing two different builds of Swift-DocC at the same time.

The `name` that's used in the base URL contains both the name of the test class and the test case.

Together this means that only tests in the same checkout and in the same test method will have the same base URL. Since the teardown block runs after the test method there's no risk that a base URL is removed while there are tests working on subpaths of that URL.

The first path component within each test base URL is `ProcessInfo.processInfo.globallyUniqueString` so two different temp directories in the same test method also wouldn't have collisions.

Reply:
> The name that's used in the base URL contains both the name of the test class and the test case.

Ah got it- based on the docs I was assuming that it was _just_ the name of the test case so I'm less concerned then.

However- the collisions I was more worried about are with other items in the bundle, for example if we added a new TestDocC target and SwiftPM emitted a new directory there to hold its output. The chances of collision seem very low though since it's both the test class and test case so I think this is okay.